### PR TITLE
kvm: use native IDT for speed-up [fixes #982]

### LINKED ISSUE #982: native IDT in KVM? (FPU performance is bad)
Here's more FPU fun. Feel free to ignore this bug, it is probably not relevant in the real world.

http://www.conradshome.com/win31/archive/softlib/winfloat.exe contains a small FPU benchmark in floatapp.exe. It reports the following run times:

dosemu2 with _mathco=off or hide87: 6500-7000ms
dosemu2 with HW FPU: ~3500 ms
wine: 5ms

All 3 are on my i7 macbook pro (cpuinfo-grey.txt from bug #976). For comparison, real Windows 7 32 bit on a 10 years older machine (Core 2 Duo) runs the benchmark in 15ms. I have no clue how fast it would be on a Pentium 1 from 1995.

**To Reproduce**
Download and unzip http://www.conradshome.com/win31/archive/softlib/winfloat.exe, run floatapp.exe in Windows 3.x, click "benchmark" in the menu.

**Attach the log**
A -D+9M log is attached. Logging DPMI calls slows down things a lot more, so I guess the performance comes from entering and leaving the DPMI guest.

[boot.log.gz](https://github.com/stsp/dosemu2/files/4027016/boot.log.gz)

FloatApp also includes an exception test that reproduces bug 976 with HW FPU, but works OK with _mathco=off. Wine apparently can't handle Win16 FPU exceptions at all and just crashes in the 64 bit ntdll exception handler.

### PR DESCRIPTION
**Pull request by bartoldeman**
This is my patch adapted to Stas' notation.

I'm still not sure about 0x70-0x77, 0x1c, 0x23, 0x24 so kept excluding them just to be cautious.

### CODE REVIEW

**Comment by stsp** [comment]
```
dpmi.c: In function ‘dpmi_set_interrupt_vector’:

dpmi.c:1566:13: error: too many arguments to function ‘kvm_set_idt_default’

             kvm_set_idt_default(num, -1, 0);

             ^~~~~~~~~~~~~~~~~~~

In file included from dpmi.c:65:0:
```

File: src/base/emu-i386/kvm.c [hunk 1]
```diff
@@ -133,6 +134,40 @@ static struct kvm_userspace_memory_region maps[MAXSLOT];
...
+static void set_idt_default(dosaddr_t mon, int i)
+{
+    unsigned int offs = mon + offsetof(struct monitor, code) + i * 16;
+    monitor->idt[i].offs_lo = offs & 0xffff;
+    monitor->idt[i].offs_hi = offs >> 16;
+    monitor->idt[i].seg = 0x8; // FLAT_CODE_SEL
+    /* DPL must be 0 so that software ints from DPMI clients will GPF.
+       Exceptions are int3 (BP) and into (OF): matching the Linux kernel
+       they must generate traps 3 and 4, and not GPF */
+    monitor->idt[i].DPL = (i == 3 || i == 4) ? 3 : 0;
```
Comment (stsp) [inline src/base/emu-i386/kvm.c]: This is probably not needed here, as
there is a "< 32" check that would exclude
this. Which is why I didn't move that line.
This is minor of course.

**Comment by stsp** [comment]
I tried to fix the compilation problems
manually (there are more than one), but
the patch doesn't work at the end.

File: src/base/emu-i386/kvm.c (hunk 1 above)
Comment (bartoldeman) [reply src/base/emu-i386/kvm.c]: yes this is minor. But you actually moved the line but not the comment :)
The check is in kvm_set_idt_default(), not this set_idt_default, which is also called from the init code. Safer to just keep it here.

**Comment by bartoldeman** [comment]
I missed installation of libreadline-dev, which is why the compilation did not include USE_MHPDBG for me... Fixed now.

I also double checked with cwsdpmi source and it looks like we are indeed safe for hw ints and 0x23/0x24: cwsdpmi needs to deal with hw ints generating int xx via the IDT and we don't. And for the special ones it's about realmode call backs for RM ints which we already have too.

**Comment by stsp** [comment]
This patch doesn't work.
Try bc.exe or win31.

**Comment by bartoldeman** [comment]
Can you try again? 16-bit IDT descriptors were not handled properly.

**Comment by stsp** [comment]
Works perfectly, thanks!

File: src/dosext/dpmi/dpmi.c [hunk 2]
```diff
@@ -1554,6 +1554,18 @@ void dpmi_set_interrupt_vector(unsigned char num, DPMI_INTDESC desc)
...
+        /* KVM: we can directly use the IDT but don't do it when debugging */
+#ifdef USE_MHPDBG
+        if (mhpdbg.active && dpmi_mhp_intxxtab[num])
+            kvm_set_idt_default(num);
+        else
+#endif
+        if (desc.selector == dpmi_sel())
+            kvm_set_idt_default(num);
+        else
+            kvm_set_idt(num, desc.selector, desc.offset32, DPMI_CLIENT.is_32);
```
Comment (stsp) [inline src/dosext/dpmi/dpmi.c]: Mm, are you sure of that?
Maybe we should check is_32 of
the handler code selector, like
`Segments[desc.selector>>3].is_32`?

Comment (bartoldeman) [reply src/dosext/dpmi/dpmi.c]: It's consistent with `dpmi_iret` this way where it was causing issues without my last change. I would need to check dpmi to see what is correct, but it needs to be consistent for sure.

Comment (bartoldeman) [reply src/dosext/dpmi/dpmi.c]: It's consistent with `dpmi_iret` this way where it was causing issues without my last change. I would need to check dpmi to see what is correct, but it needs to be consistent for sure.

Comment (bartoldeman) [reply src/dosext/dpmi/dpmi.c]: I think DPMI_CLIENT.is_32 is correct, from http://www.delorie.com/djgpp/doc/dpmi/api/310205.html: "32-bit clients must supply a 32-bit offset in the EDX register. If the client's handler chains to the next exception handler it must do so using a 32-bit interrupt stack frame."

Checking the desc.selector for 32-bit would contradict this if it's a 16-bit routine with a 32-bit stack.

**Comment by stsp** [comment]
OK, not trying to claim the opposite,
and the argument about dpmi iret is
valid too. My point was just that real `iret`
doesn't check for dpmi client, it checks
iirc the selector bitness, so we could
do that too (but don't have to if we
always trusted the client bitness before).

**Pull request merged**